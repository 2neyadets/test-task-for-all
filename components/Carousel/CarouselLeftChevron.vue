<template lang="pug">
  .absolute.flex.flex-center.carousel__left-chevron
    Btn(
      @click.native="$emit('changeSlide', currentSlide - 1)"
      icon="chevron_left"
      :disable="currentSlide === 0"
    )
</template>

<script>

import Btn from '../Reusable/Btn'
export default {
  name: 'CarouselLeftChevron',
  components: { Btn },
  props: {
    currentSlide: {
      type: Number,
      required: true,
    },
    amountOfSlides: {
      type: Number,
      required: true,
    },
  },
  data () {
    return {
    }
  },
  computed: {},
  watch: {},
  created () {},
  mounted () {},
  updated () {},
  methods: {},
}
</script>

<style scoped lang="stylus">
  .carousel__left-chevron
    top 16px
    bottom 16px
    left 12px
</style>
