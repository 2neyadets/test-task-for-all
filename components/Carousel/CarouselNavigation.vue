<template lang="pug">
  .absolute.flex.carousel__navigation
    .carousel__navigation-inner.flex.justify-center
      template(v-for="(slide, index) in amountOfSlides")
        Btn.q-mx-xs(
          @click.native="$emit('changeSlide', index)"
          icon="brightness_1"
          :size="12"
          :active="index === currentSlide"
          round
        )
</template>

<script>

import Btn from '../Reusable/Btn'
export default {
  name: 'CarouselNavigation',
  components: { Btn },
  props: {
    currentSlide: {
      type: Number,
      required: true,
    },
    amountOfSlides: {
      type: Number,
      required: true,
    },
  },
  data () {
    return {
    }
  },
  computed: {},
  watch: {},
  created () {},
  mounted () {},
  updated () {},
  beforeDestroy () {},
  methods: {},
}
</script>

<style scoped lang="stylus">
  .carousel__navigation
    left 16px
    right 16px
    bottom 16px
    overflow-x auto
    overflow-y hidden

    &-inner
      flex 1 1 auto
</style>
