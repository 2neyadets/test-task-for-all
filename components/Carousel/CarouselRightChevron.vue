<template lang="pug">
  .absolute.flex.flex-center.carousel__right-chevron
    Btn(
      @click.native="$emit('changeSlide', currentSlide + 1)"
      icon="chevron_right"
      :disable="currentSlide === amountOfSlides - 1"
    )
</template>

<script>

import Btn from '../Reusable/Btn'
export default {
  name: 'CarouselRightChevron',
  components: { Btn },
  props: {
    currentSlide: {
      type: Number,
      required: true,
    },
    amountOfSlides: {
      type: Number,
      required: true,
    },
  },
  data () {
    return {
    }
  },
  computed: {},
  watch: {},
  created () {},
  mounted () {},
  updated () {},
  methods: {},
}
</script>

<style scoped lang="stylus">
  .carousel__right-chevron
    top 16px
    bottom 16px
    right 12px
</style>
