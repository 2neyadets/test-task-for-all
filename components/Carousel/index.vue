<template lang="pug">
  .carousel.bg-transparent.text-white.shadow-12.rounded-borders.relative-position(style="min-height: 484px;")
    .carousel__wrapper.column.flex-center
      .carousel__slide
        .relative-position.row
          template(v-for="(note, index) in notesForCurrentSlide")
            .col
              Card(:note="note" :index="index" :key="note.id")
    CarouselNavigation(:currentSlide="currentSlide" @changeSlide="changeSlide" :amountOfSlides="amountOfSlides")
    CarouselLeftChevron(:currentSlide="currentSlide" @changeSlide="changeSlide" :amountOfSlides="amountOfSlides")
    CarouselRightChevron(:currentSlide="currentSlide" @changeSlide="changeSlide" :amountOfSlides="amountOfSlides")
</template>

<script>
import { mapGetters } from 'vuex'
import Card from '../Reusable/Card'
import CarouselNavigation from './CarouselNavigation'
import CarouselLeftChevron from './CarouselLeftChevron'
import CarouselRightChevron from './CarouselRightChevron'

export default {
  name: 'Carousel',
  components: { Card, CarouselRightChevron, CarouselLeftChevron, CarouselNavigation },
  data () {
    return {
      currentSlide: 0,
      lastCardsInSlide: 0,
    }
  },
  computed: {
    ...mapGetters([
      'viewportWidth',
    ]),
    ...mapGetters('notes', [
      'notesArr',
    ]),
    cardsInSlide () {
      if (this.viewportWidth >= 1024) {
        return 3
      } else if (this.viewportWidth >= 750) {
        return 2
      } else {
        return 1
      }
    },
    notesForCurrentSlide () {
      return [...this.notesArr].splice(this.currentSlide * this.cardsInSlide, this.cardsInSlide)
    },
    amountOfSlides () {
      return Math.ceil(this.notesArr.length / this.cardsInSlide)
    },
  },
  watch: {
    currentSlide (v) {
      localStorage.setItem('currentSlide', JSON.stringify(v))
    },
    cardsInSlide (v) {
      const isAmountOfCardsIncreased = this.lastCardsInSlide < v
      this.lastCardsInSlide = v
      const lastViewedNoteStartIndex = this.currentSlide * (isAmountOfCardsIncreased ? (v - 1) : (v + 1))
      console.log('lastViewedNoteStartIndex: ', lastViewedNoteStartIndex)
      for (let i = 0; i < this.amountOfSlides; i++) {
        if (lastViewedNoteStartIndex < (i + 1) * v) {
          this.currentSlide = i
          break
        }
      }
    },
  },
  created () {},
  mounted () {
    this.lastCardsInSlide = this.cardsInSlide
    const localCurrentSlide = localStorage.getItem('currentSlide') || null
    this.currentSlide = (localCurrentSlide && JSON.parse(localCurrentSlide)) || 0
  },
  updated () {
    if (this.notesForCurrentSlide.length === 0 && this.currentSlide !== 0) this.currentSlide = this.currentSlide - 1
  },
  methods: {
    changeSlide (currentSlide) {
      if (currentSlide >= 0 && currentSlide < this.amountOfSlides) this.currentSlide = currentSlide
    },
  },
}
</script>

<style scoped lang="stylus">
  .carousel

    &__wrapper
      padding 16px 50px 50px

    &__slide
      width 100%
      margin 0
      padding 0
      display flex
      justify-content center
      align-items center
      z-index 0
</style>
