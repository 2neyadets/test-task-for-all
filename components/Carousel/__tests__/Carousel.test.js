import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Carousel from '../index'
import layoutStore from '../../../store/index'
const layoutActions = layoutStore.actions,
  layoutMutations = layoutStore.mutations,
  layoutState = layoutStore.state,
  layoutGetters = layoutStore.getters
import noteStore from '../../../store/notes'
const noteActions = noteStore.actions,
  noteMutations = noteStore.mutations,
  noteState = noteStore.state,
  noteGetters = noteStore.getters

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Carousel component', () => {
  let wrapper, store
  const notesArrVar = [
    {
      id: 'f17351326a68-f48d2c4d',
      title: '11111111111111111111',
    },
    {
      id: 'f17351327bb8-f159ba6c',
      title: '22222222222222222222',
    },
    {
      id: 'f17351328cd9-f20337e2',
      title: '33333333333333333333',
    },
    {
      id: 'f17351326a68-f48d2cdd',
      title: '44444444444444444444',
    },
    {
      id: 'f17351327bb8-f159baee',
      title: '55555555555555555555',
    },
    {
      id: 'f17351328cd9-f20337ff',
      title: '66666666666666666666',
    },
  ]

  beforeEach(() => {
    store = new Vuex.Store({
      actions: layoutActions,
      mutations: layoutMutations,
      state: layoutState,
      getters: layoutGetters,
      modules: {
        notes: {
          namespaced: true,
          actions: noteActions,
          mutations: noteMutations,
          state: noteState,
          getters: noteGetters,
        },
      },
    })

    // init store
    store.dispatch('notes/changeNotes', notesArrVar)
    store.dispatch('screenChanged', { innerWidth: 1024, innerHeight: 768 })

    wrapper = shallowMount(Carousel, {
      localVue,
      store,
    })

    wrapper.setData({
      currentSlide: 0,
    })
  })

  it('mounts without errors', done => {
    localVue.config.errorHandler = done // для отлавливания ошибок внутри nextTick, если нужно что-то выпонять внутри него
    expect.assertions(1)
    expect(wrapper).toBeTruthy()
    done()
  })

  it('can change cardsInSlide', async () => {
    expect.assertions(4)
    let cards = wrapper.findAll('Card-stub')
    expect(cards.length).toBe(3)
    wrapper.setData({
      currentSlide: 1,
    })
    await localVue.nextTick()
    store.dispatch('screenChanged', { innerWidth: 800, innerHeight: 768 })
    await localVue.nextTick()
    cards = wrapper.findAll('Card-stub')
    expect(cards.length).toBe(2)
    store.dispatch('screenChanged', { innerWidth: 400, innerHeight: 768 })
    await localVue.nextTick()
    cards = wrapper.findAll('Card-stub')
    expect(cards.length).toBe(1)
    store.dispatch('screenChanged', { innerWidth: 1024, innerHeight: 768 })
    await localVue.nextTick()
    cards = wrapper.findAll('Card-stub')
    expect(cards.length).toBe(3)
  })

  it('can show correct notesForCurrentSlide', async () => {
    expect.assertions(12)
    let notesTitlesArrForFirstSlide = wrapper.vm.notesForCurrentSlide.map(note => note.title)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[0].title)).toBe(true)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[1].title)).toBe(true)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[2].title)).toBe(true)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[3].title)).toBe(false)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[4].title)).toBe(false)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[5].title)).toBe(false)
    wrapper.setData({
      currentSlide: 1,
    })
    await localVue.nextTick()
    notesTitlesArrForFirstSlide = wrapper.vm.notesForCurrentSlide.map(note => note.title)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[0].title)).toBe(false)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[1].title)).toBe(false)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[2].title)).toBe(false)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[3].title)).toBe(true)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[4].title)).toBe(true)
    expect(notesTitlesArrForFirstSlide.includes(notesArrVar[5].title)).toBe(true)
  })

  it('can handle "changeSlide" event', () => {
    expect.assertions(1)
    const rightChevron = wrapper.find('CarouselRightChevron-stub')
    rightChevron.vm.$emit('changeSlide', 2)
    expect(rightChevron.emitted('changeSlide')[0]).toEqual([2])
  })

  it('match snapshot', () => {
    expect.assertions(1)
    expect(wrapper.element).toMatchSnapshot()
  })
})
