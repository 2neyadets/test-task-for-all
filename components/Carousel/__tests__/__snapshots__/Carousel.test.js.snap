// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component match snapshot 1`] = `
<div
  class="carousel bg-transparent text-white shadow-12 rounded-borders relative-position"
  style="min-height: 484px;"
>
  <div
    class="carousel__wrapper column flex-center"
  >
    <div
      class="carousel__slide"
    >
      <div
        class="relative-position row"
      >
        <div
          class="col"
        >
          <card-stub
            index="0"
            note="[object Object]"
          />
        </div>
        <div
          class="col"
        >
          <card-stub
            index="1"
            note="[object Object]"
          />
        </div>
        <div
          class="col"
        >
          <card-stub
            index="2"
            note="[object Object]"
          />
        </div>
      </div>
    </div>
  </div>
  <carouselnavigation-stub
    amountofslides="2"
    currentslide="0"
  />
  <carouselleftchevron-stub
    amountofslides="2"
    currentslide="0"
  />
  <carouselrightchevron-stub
    amountofslides="2"
    currentslide="0"
  />
</div>
`;
